subsys/bluetooth/mesh/access.c

```diff
@@ -2614,3 +2634,220 @@ uint8_t bt_mesh_comp_parse_page(struct net_buf_simple *buf)
+	struct k_work_delayable random_delay;
+	struct access_message *pending_msg;
+} access_randomizer = {.random_delay = Z_WORK_DELAYABLE_INITIALIZER(randomizer_handler)};
+
+K_MEM_SLAB_DEFINE_STATIC(access_message_chunks, sizeof(struct access_message_chunk),
+			 CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_COUNT, 4);
+
+static void reschedule_randomizer(void)
+{
+	uint32_t fired_time_min = 0;
+	uint32_t curr_time;
+	k_timeout_t delay = K_NO_WAIT;
+
+	access_randomizer.pending_msg = NULL;
+
+	for (int i = 0; i < CONFIG_BT_MESH_DELAYED_ACCESS_MSG_COUNT; i++) {
+		if (access_messages[i].busy) {
+			if (fired_time_min < access_messages[i].fired_time) {
+				fired_time_min = access_messages[i].fired_time;
+				access_randomizer.pending_msg = &access_messages[i];
+			}
+		}
+	}
+
+	if (!access_randomizer.pending_msg) {
+		return;
+	}
+
+	curr_time = k_uptime_get_32();
+	if (curr_time < fired_time_min) {
+		delay = K_MSEC(fired_time_min - curr_time);
+	}
+
+	k_work_reschedule(&access_randomizer.random_delay, delay);
+}
+
+static int allocate_msg_chunks(struct access_message *msg, int number)
+{
+	struct access_message_chunk *chunk;
+	int err;
+
+	for (int i = 0; i < number; i++) {
+		err = k_mem_slab_alloc(&access_message_chunks, (void **)&chunk, K_NO_WAIT);
+		if (err) {
+			LOG_WRN("Unable allocate %u chunks, allocated %u", number, i);
+			return i;
+		}
+		sys_slist_append(&msg->chunks, &chunk->node);
+	}
+
+	return number;
+}
+
+static void release_msg_chunks(struct access_message *msg)
+{
+	struct access_message_chunk *chunk;
+	sys_snode_t *node;
+
+	while ((node = sys_slist_get(&msg->chunks))) {
+		chunk = CONTAINER_OF(node, struct access_message_chunk, node);
+		k_mem_slab_free(&access_message_chunks, (void *)chunk);
+	}
+}
+
+static struct access_message *get_access_msg(const struct bt_mesh_model *model)
+{
+	struct access_message *msg = NULL;
+
+	for (int i = 0; i < ARRAY_SIZE(access_messages); i++) {
+		if (!access_messages[i].busy) {
+			msg = &access_messages[i];
+		}
+
+		if (access_messages[i].model == model) {
+			msg = &access_messages[i];
+			break;
+		}
+	}
+
+	if (!msg) {
+		LOG_DBG("No available access message context. Purge pending access message.");
+		msg = access_randomizer.pending_msg;
+	}
+
+	return msg;
+}
+
+static int push_msg_from_randomizer(struct access_message *msg)
+{
+	sys_snode_t *node;
+	struct access_message_chunk *chunk;
+	size_t len = msg->len;
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(buf, len);
+
+	SYS_SLIST_FOR_EACH_NODE(&msg->chunks, node) {
+		size_t tmp = MIN(CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE, len);
+
+		chunk = CONTAINER_OF(node, struct access_message_chunk, node);
+		memcpy(net_buf_simple_pull_mem(&buf, tmp), chunk->data, tmp);
+		len -= tmp;
+	}
+
+	err = push_msg_to_transport(msg->model, &msg->ctx, &buf, msg->cb, msg->cb_data);
+	if (err != -EBUSY && err != -ENOBUFS) {
+		release_msg_chunks(msg);
+		msg->busy = false;
+		if (err && msg->cb && msg->cb->end) {
+			msg->cb->end(err, msg->cb_data);
```

Comment:
Should be start callback, not end, as the message transmission hasn't started yet.